### PR TITLE
Validate default response value received when creating/updating assets

### PR DESCRIPTION
**Pull request by WinnyTroy**
Validate values received from the Default Response field

Addresses onaio/infrastructure#3589

### CODE REVIEW

File: jsapp/js/actions.es6 [hunk 1]
```diff
@@ -375,8 +375,12 @@ actions.resources.updateAsset.listen(function(uid, values, params={}) {
         actions.resources.updateAsset.completed(asset);
         resolve(asset);
       })
-      .fail(function(...args){
-        reject(args)
+      .fail((args) => {
+        if (args.status === 500) {
+          reject(args)
+        } else {
+          alertify.error(t(args.responseText));
```
Comment (ukanga) [inline jsapp/js/actions.es6]: Should we still` reject(args)` after `alertify.error()` or is`alertify.error()` a Promise that is rejected?

**Review by moshthepitt — COMMENTED** [review]
Can we add tests please?

File: jsapp/js/actions.es6 (hunk 1 above)
Comment (WinnyTroy) [reply jsapp/js/actions.es6]: Thanks for catching this.
Alertify is not a promise. We use `alertify.error` here to generate the nice looking browser pop ups to showcase the error to the user.
I called reject to complete the async operation after notifying the user.

File: kpi/models/asset.py [hunk 2]
```diff
@@ -390,6 +392,16 @@ def _rename_translation(self, content, _from, _to):
             raise ValueError('Duplicate translation: {}'.format(_to))
         _ts[_ts.index(_from)] = _to
 
+    def _contains_invalid_chars(self, content):
+        for row in content['survey']:
+            try:
+                if row['default'] and bool(re.search(
+                        r'[<|>|&]', row['default'])):
+                    raise ValidationError(
+                        'XForm questions settings may contain malicious content')
```
Comment (moshthepitt) [inline kpi/models/asset.py:401]: Do we usually get such strings from a constants/strings module/file?

File: kpi/tests/test_assets.py [hunk 3]
```diff
@@ -432,6 +432,18 @@ def test_templates_move_form_title_to_name(self):
...
+        unsanitized_html_str = "<h1>HTML Injection testing</h1>"
+        _content = self._content()
+        _content['survey'][0]['default'] = unsanitized_html_str
+
+        def _create_asset():
+            Asset.objects.create(
+                content=_content,
+                owner=self.user,
+                asset_type='survey')
+        self.assertRaises(ValidationError, _create_asset)
```
Comment (moshthepitt) [inline kpi/tests/test_assets.py:445]: 1. Should we assert that the validation error message is as expected?
2. Is the case where an error is not raised already tested?

Comment (WinnyTroy) [reply kpi/tests/test_assets.py:445]: Ive added these test scenerios

File: kpi/models/asset.py (hunk 2 above)
Comment (WinnyTroy) [reply kpi/models/asset.py:401]: I dont believe so. The error message string is passed to the Exception.

Comment (moshthepitt) [reply kpi/models/asset.py:401]: What I was asking was whether this project has something similar to `common_tags` (onadata.libs.utils.common_tags) in Onadata and whether the convention is to use it or not.

**Comment by moshthepitt** [comment]
@DavisRayM @WinnyTroy Non-blocking question: How do we plan to keep this repo up-to-date with the upstream one?  Are we going to try and get our changes merged there?

cc: @ukanga @denniswambua

**Comment by faith-mutua** [comment]
@WinnyTroy the vulnerability issue has been resolved. The error in the attached screenshot is being thrown when saving a form that has malicious content. The form is never saved.
Screenshot 2021-03-05 at 12 37 11

File: kpi/serializers/v2/asset.py [hunk 4]
```diff
@@ -135,7 +136,10 @@ def update(self, asset, validated_data):
             except ValueError as err:
                 raise serializers.ValidationError(str(err))
             validated_data['content'] = asset_content
-        return super().update(asset, validated_data)
+        try:
+            return super().update(asset, validated_data)
+        except ValidationError as e:
+            raise serializers.ValidationError(e.message)
```
Comment (moshthepitt) [inline kpi/serializers/v2/asset.py]: I think the accepted thing to do these days would be something like `raise serializers.ValidationError(e.message) from e`.  Here's the [documentation](https://docs.python.org/3/library/exceptions.html#built-in-exceptions).

Comment (WinnyTroy) [reply kpi/serializers/v2/asset.py]: Thanks for this.
I've made that change

File: kpi/models/asset.py (hunk 2 above)
Comment (WinnyTroy) [reply kpi/models/asset.py:401]: No, unfortunately I have not found anything similar to Onadata's `common_tag.py` file.
The repository does have a `constants.py` file, but from the looks of it, it wouldn't be a correct fit to place this string value.

**Review by ukanga — APPROVED** [review]

**Review by DavisRayM — APPROVED** [review]
LGTM

**Pull request merged**